[web_server] Adds the ability to handle Private Network Access preflight requests

DanielBaulig opened the pull request:
# What does this implement/fix?

Amends the web_server component to be able to correctly handle and respond to Private Network Access preflight requests. Resolves the feature request https://github.com/esphome/feature-requests/issues/2453

Private Network Access is a new feature expecting to land in Chrome 123 which provides secure (HTTPS/TLS) web applications the ability to interact with devices on private networks, even if those are not hosted securely. 

This has previously been impossible due to the browsers mixed content policy preventing a secure web application from loading any resources or making any requests to non-secure web sites, including those on private networks, which due to lack of access to the DNS/PKI infratructure typically cannot be hosted using TLS/HTTPS.

The Private Network Access specification defines a flow in which a user can grant the web application explicit permission to access specific private network origins. The specification requires the target HTTP server to correctly handle and respond to OPTIONS preflight requests to trigger the permission dialog in the first place and allow subsequent requests.

The new API is currently in Origin Trial and can be accessed and enabled in Chrome Nightly. More about it can be learned in [the specification](https://wicg.github.io/private-network-access/), [this explainer](https://docs.google.com/document/d/1W70cFFaBGWd0EeOOMxJh9zkmxZ903vKUaGjyF-w7HcY/edit) and over on this [Github issue tracker](https://github.com/WICG/private-network-access/issues/117).


## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/2453

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/3338

## Test Environment
- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
```yaml
# Example config.yaml
web_server:
```
Note: [Chrome Canary](https://www.google.com/chrome/canary/) with enabled Private Network Access Permission Prompt (see about:flags) is also required. From there, after flashing an ESP device with the updated web_server component, open any secure website (e.g. https://esphome.io) in Chrome Canary, open the Developer Tools and use the following request code to send a private network access request to the ESP device: `fetch('http://esp-device-hostname.local', { targetAddressSpace: 'private' })`. The permission prompt should show up and give you the option to allow this (and subsequent) requests from the same origin. Look at the Network tab to see an OPTIONS preflight request to the ESP web server that was confirmed with a 200. Depending on if you have granted Private Network Access or not using the permission dialog there will be either an additional successful or a failed GET fetch request to the ESP device.

Repeat the same experiment with either the Private Network Access Permission Prompt flag disabled or on a stable release channel of Chrome and you will see that the request to the ESPHome device will always fail. Repeat the experiment with an unpatched version of the ESPHome web_server and you will see that the OPTIONS preflight request will always fail and no permission dialog will be shown.

Note: to allow for POST requests (such as those that will trigger state changes) may require additional changes. In particular I believe it may also require presence of `Access-Control-Allow-Methods: *` or the more restrictive `Access-Control-Allow-Methods: GET, POST`. Happy to do some more investigation into this and provide another pull request for that as well.

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
          N/A
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

DanielBaulig committed:
[web_server] Adds the ability to handle Private Network Access preflight requests

Resolves the feature request https://github.com/esphome/feature-requests/issues/2453

DanielBaulig committed:
Remove helpers include

nagyrobi commented:
This needs to be inplemented in idf too.

DanielBaulig commented:
> This needs to be inplemented in idf too.

I am not familiar with how the IDF web server relates to the web_server component. Are you saying this needs to be implemented in the IDF server before it could be landed (e.g. because they must maintain feature parity or something like that) or are you saying it would be nice to have this feature in the IDF server, too, but fundamentally that is not something blocking this change?

nagyrobi commented:
Please see the doc: https://esphome.io/components/esp32.html#esp-idf-framework

ESPHome runs many conponents in

DanielBaulig commented:
Thank you for sharing that link. I'm afraid I'm still unsure about what the answer to my question is:

> you are asking that I make sure that the new feature also works when building using the IDF framework over the Arduino framework. Is that correct?

If that's the case, happy to do so. I'm currently trying to get the build working with the IDF framework at all. Running into some (unrelated) build errors.

nagyrobi commented:
Yes,please make sure you implement this in both Arduino and ESP-IDF frameworks.

DanielBaulig commented:
Thank you for sharing that link. That makes a lot of sense.

Let me see what needs to be done to make this work within the IDF framework, too. I'm currently trying to get the build working with the IDF framework. Right now, it's not building at all for me (unrelated to my changes).

nagyrobi commented:
And you need to make a docs PR too for the maintainers to review the whole thing.

DanielBaulig commented:
> And you need to make a docs PR too for the maintainers to review the whole thing.

As I mentioned in my PR, I will do so once I am getting a positive signal that this PR is something that is considered to be merged.

nagyrobi commented:
It's the other way around, dude...

DanielBaulig commented:
I appreciate your concern for me getting these changes reviewed. This isn't my first contribution to ESPHome and I believe I understand how the process works. I would like to make sure that this change is something -- on a fundamental level -- that the project is willing to take, before I go an polish it to meet the bar for a review and merge. I didn't have any problems with this before and personally don't think that's an unreasonable approach. If that is what is going to block getting feedback from maintainers, then I will move forward with it at an appropriate time.

I would like to point out that your choice of words comes off as a little hostile and aggressive. I recognize that you've been helping me so far and that this might not be your intention. If so, please let me know and I will adjust my reading here. That said, I don't think calling strangers "dude" is something that is generally understood to be a friendly tone.

DanielBaulig commented:
Also, I am still running into issues building using esp-idf. I decided to open [an issue here](https://github.com/esphome/issues/issues/5070). I am not sure if this is something I should expect or even a specific problem on my end. I've so far not used esp-idf an always built my images using the Arduino framework.

jesserockz commented:
I have not reviewed yet except to understand what it is for, but yes this something we will want merged when ready. Please continue and make the docs PR =)

DanielBaulig commented:
Will do. Appreciate it, thanks!

What's the stance on providing esp-idf support with this PR? I attempted to look into it given the feedback i have received here. But esp-idf appears to not build *at all* in my environment.

Will ping you when it's ready to be reviewed.

jesserockz commented:
For your idf build issues, I am not sure whats going on there, seems like the idf auto-created virtualenv is having issues. 
Building with docker on my computer works as expected.

Are you not installing esphome on your local computer and compiling that way? It will always be faster than a raspberry pi. 

But lets also not fill this PR up with "unrelated" issues =)

DanielBaulig commented:
Sorry, didn’t mean to discuss the build issue per se.My question was more around if having a working ESP-IDF implementation of this feature is going to be a merge blocker or if an Arduino implementation could be accepted by itself?On Nov 5, 2023, at 2:01 PM, Jesse Hills ***@***.***> wrote:﻿
For your idf build issues, I am not sure whats going on there, seems like the idf auto-created virtualenv is having issues.
Building with docker on my computer works as expected.
Are you not installing esphome on your local computer and compiling that way? It will always be faster than a raspberry pi.
But lets also not fill this PR up with "unrelated" issues =)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

jesserockz commented:
Ah I see, new features for components should work with all supported platforms for that component.

DanielBaulig committed:
TEMP

DanielBaulig commented:
I setup a development environment and Docker locally and got esp-idf to compile on Apple silicon. I am looking at how ESP-IDF uses `web_server` and how the `web_server` and `web_server_idf` components interact with each other. I assume there are a lot of practical differences between them, but what stands out immediately is that the Arduino version of `AsyncWebServerRequest` (I believe implemented in `ESPAsyncWebServer`) exposes a method `getHeader` whereas the ESP-IDF version of `AsyncWebServerRequest` (implemented inside of ESPHome's `web_server_idf` component) exposes a method `get_header` instead. In addition, some other methods (e.g. `hasHeader`, `addInterestingHeader`) that I'm using in the new code don't have any equivalents in `web_server_idf`'s version of `AsyncWebServerRequest` right now.

Is the idea that `AsyncWebServerRequest` in `web_server_idf` functions as kind of a compatibility/adapter layer between Arduino `AsyncWebServer` (and `AsyncWebServerRequest`, etc) and the ESP-IDF APIs? If so, I assume the preferred way to proceed would be to make `web_server_idf`'s implementation of `AsyncWebServerRequest` API compatible with the Arduino `AsyncWebServerRequest` implementation? E.g. rename `get_header` method to `getHeader` (or add a version thereof that is compatible in signature to Arduino's `getHeader`) and add `hasHeader` and `addInterestingHeader` implementations to `web_server_idf`'s version of `AsyncWebServerRequest`?

Or should I instead avoid that and try to work as much as possible/needed with preprocessor branches (`#ifdef USE_ARDUINO` / `#ifdef USE_ESP_IDF`) inside of `web_server` to make the code work with both Arduino and ESP-IDF?
Some guidance would be greatly appreciated.

DanielBaulig committed:
Switch handler order

DanielBaulig committed:
Add esp idf support

DanielBaulig committed:
Fix some formatting and lint issues

DanielBaulig commented:
Added documentation PR https://github.com/esphome/esphome-docs/pull/3338

jesserockz commented:
As the _idf component provides a compatability layer, yes perhaps it should also expose `getHeader`

DanielBaulig committed:
Merge branch 'dev' into web_server_pna

DanielBaulig committed:
Fix clang-format lints

DanielBaulig commented:
I added a `hasHeader` equivalent in `web_server_idf`. Since I am not using `getHeader`/`get_header` (and wouldn't have a great way to test it) I decided to leave that method alone for now, even though it should probably be renamed to `getHeader` in `web_server_idf`.

Also instead of adding an empty `addInterestingHeader` to `web_server_idf`, I decided to guard the call in `web_server` with a preprocessor condition. It seemed cleaner, since this a call that is not required using the IDF framework APIs. I also added support for `DefaultHeaders` into the EventSource code in `web_server_idf`, which was missing and is required for this to work. There may be an opportunity to make `web_server_idf`'s EventSource piggy-back on `AsyncWebServerRequest` somehow to avoid this kind of logic duplication, but since that seems to be a somewhat more complex undertaking I decided to not make that change in the context of this PR.

I believe I also fixed all the lint and formatting issues.

I think this should be ready for review now.

DanielBaulig commented on esphome/components/web_server/web_server.cpp:
```diff
@@ -359,6 +364,17 @@ void WebServer::handle_index_request(AsyncWebServerRequest *request) {
+  AsyncWebServerResponse *response = request->beginResponse(200, "");
+  response->addHeader(HEADER_CORS_ALLOW_PNA, "true");
+  response->addHeader(HEADER_PNA_NAME, App.get_name().c_str());
+  std::string mac = get_mac_address_pretty();
+  response->addHeader(HEADER_PNA_ID, mac.c_str());
+  request->send(response);
```
I added the local `mac` variable here because I had some issues with the lifetime of the string in the IDF framework otherwise and the header would sometimes not contain the correct string when the response was received by the browser. Storing the string in a local variable resolved the issue.

DanielBaulig commented on esphome/components/web_server/web_server.cpp:
```diff
@@ -34,6 +34,11 @@ namespace web_server {
 
 static const char *const TAG = "web_server";
 
+static const char *const HEADER_PNA_NAME = "Private-Network-Access-Name";
+static const char *const HEADER_PNA_ID = "Private-Network-Access-ID";
+static const char *const HEADER_CORS_REQ_PNA = "Access-Control-Request-Private-Network";
+static const char *const HEADER_CORS_ALLOW_PNA = "Access-Control-Allow-Private-Network";
+
```
Should probably guard those with a preprocessor condition, too. Will fix.

DanielBaulig committed:
Guard header strings with preprocessor condition

DanielBaulig committed:
Ignore lint for Arduino compatible methods

DanielBaulig committed:
Fix whitespace

DanielBaulig committed:
Merge branch 'dev' into web_server_pna

DanielBaulig commented:
Thanks @jesserockz!

jesserockz commented:
No, Thank you!
